src/tactic/rewrite_search/explain.lean

```diff
@@ -0,0 +1,255 @@
+-/
+
+import tactic.rewrite_search.types
+import tactic.converter.interactive -- Required for us to emit more compact `conv` invocations
+
+/-!
+# Tools to extract valid Lean code from a path found by rewrite search.
+-/
+
+open interactive interactive.types expr tactic
+
+namespace tactic.rewrite_search
+
+universes u
+
+/--
+A `dir_pair` is a pair of items designed to be accessed according to
+`dir`, a "direction" defined in the `expr_lens` library.
+-/
+@[derive inhabited]
+structure dir_pair (α : Type u) :=
+(l r : α)
+
+namespace dir_pair
+open expr_lens
+
+variables {α β : Type} (p : dir_pair α)
+
+/-- Get one side of the pair, picking the side according to the direction. -/
+def get : dir → α
+| dir.F := p.l
+| dir.A := p.r
+
+/-- Set one side of the pair, picking the side according to the direction. -/
+def set : dir → α → dir_pair α
+| dir.F v := ⟨v, p.r⟩
+| dir.A v := ⟨p.l, v⟩
+
+/-- Convert the pair to a list of its elements. -/
+def to_list : list α := [p.l, p.r]
+
+/-- Convert the pair to a readable string format. -/
+def to_string [has_to_string α] (p : dir_pair α) : string :=
+  to_string p.l ++ "-" ++ to_string p.r
+instance has_to_string [has_to_string α] : has_to_string (dir_pair α) := ⟨to_string⟩
+
+end dir_pair
+
+/- Helper for getting the nth item in a list of rules -/
+private meta def nth_rule (rs : list (expr × bool)) (i : ℕ) : expr × bool := (rs.nth i).iget
+
+/- Convert a rule into the string of Lean code used to refer to this rule. -/
+private meta def pp_rule (r : expr × bool) : tactic string :=
+  do pp ← pp r.1, return $ (if r.2 then "←" else "") ++ (to_string pp)
+
+private meta def how.to_rewrite (rs : list (expr × bool)) : how → option (expr × bool)
+| h := nth_rule rs h.rule_index
+
+/- Explain a single rewrite using `nth_rewrite`. -/
+private meta def explain_using_location (rs : list (expr × bool)) (s : side) :
+how → tactic (option string)
+| h := do
+  rule ← pp_rule $ nth_rule rs h.rule_index,
+  return $ some ("nth_rewrite_" ++ s.to_xhs ++ " " ++ to_string h.location ++ " " ++ rule)
+
+/- Explain a list of rewrites using `nth_rewrite`. -/
+private meta def using_location.explain_rewrites (rs : list (expr × bool)) (s : side)
+(steps : list how) : tactic string := do
+  rules ← steps.mmap $ λ h : how, option.to_list <$> explain_using_location rs s h,
+  return $ string.intercalate ",\n" rules.join
+
+namespace using_conv
+
+/- `app_addr` represents a tree structure that `conv` tactics use for a rewrite. -/
+inductive app_addr
+| node (children : dir_pair (option app_addr)) : app_addr
+| rw : list ℕ → app_addr
+
+open app_addr
+
+private meta def app_addr.to_string : app_addr → string
+| (node c) := "(node " ++ ((c.to_list.filter_map id).map app_addr.to_string).to_string ++ ")"
+| (rw rws) := "(rw " ++ rws.to_string ++ ")"
+
+/-- A data structure for the result of a splice operation. -/
+@[derive inhabited]
+inductive splice_result
+-- There was more of the addr to be added left, but we hit a rw
```

Comment:
I think it makes sense to add these comments to the docstring, so that it's easy to read up on them in the hover text.

Reply:
OK